### PR TITLE
remove-background version 0.2.0

### PR DESCRIPTION
**Pull request by sjfleming**
An intermediate step on our way to more-finalized code which will be published in an upcoming paper, this PR represents a large change for `remove-background`, and this version is actively being used in many ongoing research projects.

Major changes from `v0.1.0`:
- The model has been modified so that cell counts are drawn from a negative binomial and noise counts are Poisson-distributed. Observed count data is then explicitly the sum of cell counts and noise counts.
- The output count matrix is constructed by posterior sampling the real cell counts (which are marginalized over during training).  This guarantees that the output counts are strictly less than the input counts for each entry in the count matrix.
- A new parameter `--fpr` now controls the tradeoff between signal and noise that is implicit in any noise-removal procedure. Posterior regularization is used to generate an output count matrix that targets this "false positive rate" specified by the user.  A "false positive" in this context is a real cell count that is erroneously removed.  FPR of 0.01 means that the tool will remove as much noise as possible while controlling the removal of real signal to be no more than 1% in expectation.

Many other implementation details have been modified as well.

### CODE REVIEW

**Comment by mbabadi** [comment]
@sjfleming could you squish and rebase before I review? I'd be totally fine with a giant commit, but feel free to cherrypick and break it up into several commits if you think it makes sense.

**Comment by mbabadi** [comment]
@sjfleming actually, nvm -- i can do it on github

**Comment by mbabadi** [comment]
@sjfleming mmmm nah, it doesn't work like that... you should do it yourself

File: cellbender/remove_background/train.py [hunk 1]
```diff
@@ -185,29 +210,15 @@ def run_inference(dataset_obj: SingleCellRNACountsDataset,
...
-
-    if args.model == "simple":
-
-        # If using the simple model, there is no need for p.
-        encoder = CompositeEncoder({'z': encoder_z, 'd_loc': encoder_d})
-
-    else:
+    encoder_other = EncodeNonZLatents(n_genes=count_matrix.shape[1],
+                                      z_dim=args.z_dim,
+                                      hidden_dims=[100, 50],
```
Comment (mbabadi) [inline cellbender/remove_background/train.py:215]: magic number => consts (or expose to CLI)

File: cellbender/remove_background/vae/encoder.py [hunk 2]
```diff
@@ -107,184 +128,207 @@ def forward(self, x: torch.Tensor, **kwargs) -> Dict[str, torch.Tensor]:
...
+        # TODO: is this helpful?
+        if self.x_scaling is None:
+            n_std_est = 10
+            num = int(self.n_genes * 0.4)
+            std_estimates = torch.zeros([n_std_est])
+            for i in range(n_std_est):
+                idx = torch.randperm(x.nelement())
+                std_estimates[i] = x.view(-1)[idx][:num].std().item()
+            robust_std = std_estimates.median().item()
+            self.x_scaling = 1. / robust_std / 100.  # Get values on a level field
```
Comment (mbabadi) [inline cellbender/remove_background/vae/encoder.py:283]: these confuse me :p use parentheses?

File: cellbender/remove_background/vae/encoder.py [hunk 3]
```diff
@@ -107,184 +128,207 @@ def forward(self, x: torch.Tensor, **kwargs) -> Dict[str, torch.Tensor]:
...
+        probability is a sigmoid.
+        Several heuristics are used to try to encourage a good initialization.
 
     """
 
-    def __init__(self, input_dim: int, hidden_dims: List[int], output_dim: int,
-                 log_count_crossover: float, input_transform: str = None):
-        super(EncodeNonEmptyDropletLogitProb, self).__init__()
-        self.input_dim = input_dim
+    def __init__(self,
```
Comment (mbabadi) [inline cellbender/remove_background/vae/encoder.py:181]: please refactor all weight initializations out of the `__init__` and to a private method called by `__init__`.

File: cellbender/remove_background/vae/encoder.py [hunk 4]
```diff
@@ -107,184 +128,207 @@ def forward(self, x: torch.Tensor, **kwargs) -> Dict[str, torch.Tensor]:
...
+                self.overlap_mean = (overlap.max() + overlap.min()) / 2
+                self.overlap_std = overlap.max() - overlap.min()
+            overlap = (overlap - self.overlap_mean) / self.overlap_std * 5
+        else:
+            overlap = torch.zeros_like(counts)
+
+        # Apply transformation to data.
         x = transform_input(x, self.transform)
 
+        # TODO: is this helpful?
```
Comment (mbabadi) [inline cellbender/remove_background/vae/encoder.py:274]: A good rule of thumb is to either accompany a `TODO` with github issue number, or not have it at all.

**Review by mbabadi — CHANGES_REQUESTED** [review]
Very nice code as usual :-) made a few mostly stylistic comments here and there.

**Review by mbabadi — APPROVED** [review]

**Review by mbabadi — APPROVED** [review]

**Comment by mbabadi** [comment]
@sjfleming back to you! :-)

File: cellbender/remove_background/consts.py [hunk 5]
```diff
@@ -21,10 +22,17 @@
...
-FRACTION_EMPTIES = 0.5
\ No newline at end of file
+FRACTION_EMPTIES = 0.5
+
+# Prior on rho, the swapping fraction: the two concentration parameters alpha and beta.
+RHO_ALPHA_PRIOR = 18.
+RHO_BETA_PRIOR = 200.
+
+# Prior on epsilon, the RT efficiency concentration parameter [Gamma(alpha, alpha)].
+EPSILON_PRIOR = 500.
```
[the remark this answers is not in the document]
Comment (sjfleming) [reply cellbender/remove_background/consts.py:38]: I think that is a great idea, yes, I am experimenting with that now and I think it should be part of `0.2.1`.  I do think it makes sense mechanistically that epsilon is responsible for the dispersion in the empty droplet UMI counts, rather than differences in physical droplet size (which I think is pretty tightly controlled by the instrument).

File: cellbender/remove_background/data/dataprep.py [hunk 6]
```diff
@@ -47,15 +47,18 @@ class DataLoader:
 
     def __init__(self,
                  dataset: sp.csr_matrix,
-                 empty_drop_dataset: sp.csr_matrix,
+                 empty_drop_dataset: Union[sp.csr_matrix, None],
```
[the remark this answers is not in the document]
Comment (sjfleming) [reply cellbender/remove_background/data/dataprep.py:50]: Excellent, thanks, I will make this change wherever I see it... I think I've got it in multiple places.

File: cellbender/remove_background/data/dataset.py [hunk 7]
```diff
@@ -59,19 +67,24 @@ def __init__(self,
                  fraction_empties: float = 0.5,
```
[the remark this answers is not in the document]
Comment (sjfleming) [reply cellbender/remove_background/data/dataset.py:67]: Agreed, I will remove these defaults which are always intended to be overridden.

File: cellbender/remove_background/distributions/NegativeBinomialPoissonConv.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        return out
+
+
+# We wrap the Torch distribution inside a Pyro distribution.
+# This is as simple as inheriting
+# distributions.torch_distribution.TorchDistributionMixin.
+# It adds the required extra attributes.
+class NegativeBinomialPoissonConv(TorchNegativeBinomialPoissonConv,
+                                  dist.torch_distribution
+                                  .TorchDistributionMixin):
```
[the remark this answers is not in the document]
Comment (sjfleming) [reply cellbender/remove_background/distributions/NegativeBinomialPoissonConv.py:172]: Will remove

File: cellbender/remove_background/model.py [hunk 9]
```diff
@@ -413,263 +462,76 @@ def guide(self, x: torch.Tensor):
...
             # Code specific to models with empty droplets.
             if self.include_empties:
 
-                # Sample d_empty, which doesn't depend on y.
-                pyro.sample("d_empty",
-                            dist.LogNormal(d_empty_loc,
-                                           d_empty_scale)
-                            .expand_by([x.size(0)]))
+                # Regularize based on wanting a balanced p_y_logit.
+                pyro.sample("p_logit_reg", dist.Normal(loc=enc['p_y'], scale=2.))
```
[the remark this answers is not in the document]
Comment (sjfleming) [reply cellbender/remove_background/model.py:488]: Will do

File: cellbender/remove_background/train.py (hunk 1 above)
Comment (sjfleming) [reply cellbender/remove_background/train.py:215]: Will put it in consts, as it really shouldn't need to be modified by users

File: cellbender/remove_background/vae/encoder.py (hunk 4 above)
Comment (sjfleming) [reply cellbender/remove_background/vae/encoder.py:274]: Noted, and I see I missed this one... I meant to remove it

File: cellbender/remove_background/vae/encoder.py (hunk 3 above)
Comment (sjfleming) [reply cellbender/remove_background/vae/encoder.py:181]: Sounds good

File: cellbender/remove_background/consts.py [hunk 10]
```diff
@@ -21,10 +22,17 @@
...
 # Size of minibatch by default.
 DEFAULT_BATCH_SIZE = 128
 
 # Fraction of totally empty droplets that makes up each minibatch, by default.
-FRACTION_EMPTIES = 0.5
\ No newline at end of file
+FRACTION_EMPTIES = 0.5
+
+# Prior on rho, the swapping fraction: the two concentration parameters alpha and beta.
+RHO_ALPHA_PRIOR = 18.
```
[the remark this answers is not in the document]
Comment (sjfleming) [reply cellbender/remove_background/consts.py:34]: Admittedly these are rather arbitrary and need to be revisited for `0.2.1`... the thinking was that around 8% swapping was close to the largest rate we'd typically see [18/(18+200)] (again, I'm lacking a real justification for that number), and that the relatively tight prior would encourage the model to explain as much noise as possible via swapping.

[the remark this answers is not in the document]
Comment (mbabadi) [reply cellbender/remove_background/consts.py:34]: I see -- please make a github issue about it so we won't forget to take care of it later.

File: cellbender/remove_background/consts.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mbabadi) [reply cellbender/remove_background/consts.py:38]: please make a github issue about it so we won't forget to take care of it later.

File: cellbender/remove_background/model.py [hunk 11]
```diff
@@ -115,139 +121,90 @@ def __init__(self,
                                           f"should sum to 1, but it sums to " \
                                           f"{chi_ambient_sum}"
             chi_bar_sum = np.round(dataset_obj.priors['chi_bar'].sum().item(),
```
[the remark this answers is not in the document]
Comment (sjfleming) [reply cellbender/remove_background/model.py:123]: Will make that change

File: cellbender/remove_background/model.py [hunk 12]
```diff
@@ -307,43 +261,132 @@ def model(self, x: torch.Tensor):
...
-                                      dist.LogNormal(self.d_empty_loc_prior,
-                                                     self.d_empty_scale_prior)
+                                      dist.LogNormal(loc=self.d_empty_loc_prior,
+                                                     scale=self.d_empty_scale_prior)
                                       .expand_by([x.size(0)]))
 
                 # Sample y, the presence of a real cell, based on p_logit_prior.
                 y = pyro.sample("y",
-                                dist.Bernoulli(logits=self.p_logit_prior)
+                                dist.Bernoulli(logits=self.p_logit_prior - 100.)  # TODO
```
[the remark this answers is not in the document]
Comment (sjfleming) [reply cellbender/remove_background/model.py:281]: Hmmm yes, this is a rather heuristic approach to fighting off the tendency of the solution to try to explain everything as a "cell" due to the ability of the autoencoder to explain the noise.

[the remark this answers is not in the document]
Comment (sjfleming) [reply cellbender/remove_background/model.py:281]: I look forward to a more satisfactory solution in `v0.2.1`...

File: cellbender/remove_background/model.py [hunk 13]
```diff
@@ -307,43 +261,132 @@ def model(self, x: torch.Tensor):
...
+                                            y=y,
+                                            d_cell=d_cell,
+                                            rho=rho,
+                                            chi_bar=self.avg_gene_expression)
+            else:
+                lam = torch.zeros([self.n_genes]).to(self.device)
+
+            alpha = 1. / phi
+
+            USE_NBPC = False
```
[the remark this answers is not in the document]
Comment (sjfleming) [reply cellbender/remove_background/model.py:310]: Made an attribute called `exact_log_prob`, which draws its default **init** value from consts

File: cellbender/remove_background/distributions/NegativeBinomialPoissonSumSparse.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+import torch.nn.functional as F
+from torch.distributions import constraints
+from torch.distributions.distribution import Distribution
+from torch.distributions.utils import \
+    broadcast_all, probs_to_logits, lazy_property, logits_to_probs
+
+from numbers import Number
+
+
+class TorchNegativeBinomialPoissonSumSparse(Distribution):
```
[the remark this answers is not in the document]
Comment (sjfleming) [reply cellbender/remove_background/distributions/NegativeBinomialPoissonSumSparse.py:13]: Actually, I'm going to remove this file.  It's an old version of something that we don't actually use.

**Comment by sjfleming** [comment]
@mbabadi back to you!

File: cellbender/remove_background/model.py [hunk 15]
```diff
@@ -305,18 +311,16 @@ def model(self, x: torch.Tensor):
...
-
-            USE_NBPC = False
+            alpha = phi.reciprocal()
 
-            if USE_NBPC:
+            if self.use_exact_log_prob:
 
                 # Sample gene expression from our Negative Binomial Poisson
                 # Convolution distribution, and compare with observed data.
                 c = pyro.sample("obs", NBPC(mu=mu_cell + 1e-10,
```
Comment (mbabadi) [inline cellbender/remove_background/model.py:320]: 1e-10 => consts (e.g. NPBC_MU_EPS_SAFEGAURD)
1e-10 => consts (e.g. NPBC_ALPHA_EPS_SAFEGAURD)
1e-10 => consts (e.g. NPBC_LAM_EPS_SAFEGAURD)

**Review by mbabadi — APPROVED** [review]

**Comment by mbabadi** [comment]
@sjfleming everything is looking good! just made a small comment to move the 1e-10 safeguarding shifts in NPBC to consts;  I have been bitten by various "small" epsilons in the past too many times -- it's good to have all of those in consts, just as a reminder that you are shifting things.  Ideally, you want to make sure these shifts do not affect the results within a given accepted error tolerance. Let's make an issue to further investigate NBPC approximation edge cases. If we decided to move to to a pure Poisson model, we can close that issue.

If internal tests are passing, feel free to rebase and merge (on github) :-)

File: cellbender/remove_background/model.py (hunk 15 above)
Comment (sjfleming) [reply cellbender/remove_background/model.py:320]: Will do
